bookkeeper-server/src/test/java/org/apache/bookkeeper/client/api/BookKeeperApiTest.java

```diff
@@ -0,0 +1,300 @@
+                .withAckQuorumSize(1)
+                .withWriteQuorumSize(2)
+                .withEnsembleSize(3)
+                .withDigestType(DigestType.MAC)
+                .withCustomMetadata(customMetadata)
+                .withPassword(password)
+                .execute());) {
+            lId = writer.getId();
+
+            writer.append(ByteBuffer.wrap(data));
+            writer.append(ByteBuffer.wrap(data));
+            writer.append(ByteBuffer.wrap(data)).get();
+        }
+
+        try (ReadHandle reader
+            = result(newOpenLedgerOp()
+                .withDigestType(DigestType.MAC)
+                .withPassword("bad-password".getBytes(StandardCharsets.UTF_8))
+                .withLedgerId(lId)
+                .execute())) {
+            fail("should not open ledger, bad password");
+        } catch (BKUnauthorizedAccessException ok) {
+        }
+
+        try (ReadHandle reader = result(newOpenLedgerOp()
+            .withDigestType(DigestType.CRC32)
+            .withPassword(password)
+            .withLedgerId(lId)
+            .execute())) {
+            fail("should not open ledger, bad digest");
+        } catch (BKDigestMatchException ok) {
+        }
+
+        try {
+            result(newOpenLedgerOp().execute());
+            fail("should not open ledger, no id");
+        } catch (BKNoSuchLedgerExistsException ok) {
+        }
+
+        try {
+            result(newOpenLedgerOp().withLedgerId(Long.MAX_VALUE - 1).execute());
+            fail("should not open ledger, bad id");
+        } catch (BKNoSuchLedgerExistsException ok) {
+        }
+
+        registerMockEntryForRead(lId, BookieProtocol.LAST_ADD_CONFIRMED, password, data, -1);
+        registerMockEntryForRead(lId, 0, password, data, -1);
+        registerMockEntryForRead(lId, 1, password, data, 0);
+        registerMockEntryForRead(lId, 2, password, data, 1);
+
+        try (ReadHandle reader = result(newOpenLedgerOp()
+            .withDigestType(DigestType.MAC)
+            .withPassword(password)
+            .withRecovery(false)
+            .withLedgerId(lId)
+            .execute())) {
+            assertEquals(2, reader.getLastAddConfirmed());
+            assertEquals(2, reader.readLastAddConfirmed().get().intValue());
+            assertEquals(2, reader.tryReadLastAddConfirmed().get().intValue());
+
+            checkEntries(reader.read(0, reader.getLastAddConfirmed()).get(), data);
+            checkEntries(reader.readUnconfirmed(0, reader.getLastAddConfirmed()).get(), data);
+        }
+    }
+
+    @Test
+    public void testOpenLedgerWithRecovery() throws Exception {
+        long lId;
+        byte[] data = "foo".getBytes(StandardCharsets.UTF_8);
+        final byte[] password = "password".getBytes(StandardCharsets.UTF_8);
+        Map<String, byte[]> customMetadata = new HashMap<>();
+        customMetadata.put("test", "test".getBytes(StandardCharsets.UTF_8));
+        try (WriteHandle writer = result(newCreateLedgerOp()
```

Comment:
why do you need writehandle in this test case? registerMockEntryForRead already provides entries for reading, no?